---
title: "Class 17: Biological Network Analysis"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Set up Cytoscape and R connection

Load the required packages **RCy3** from bioconductor and **igraph** from CRAN

First, we need to go in the console and install the package.
Biocmanager::install("RCy3")
```{r}
library (RCy3)
library(igraph)
```

Check our connection to cytoscape

```{r}
cytoscapePing()
```

Check the version

```{r}
cytoscapeVersionInfo()
```

We can test things further by making a small network (here in igraph format as used by the R igraph package) and sending it to Cytoscape.

```{r}
g <- makeSimpleIgraph()
createNetworkFromIgraph(g,"myGraph")
```

```{r}
plot(g)
```

```{r}
degree(g)
```


```{r}
fig <- exportImage(filename="demo", type="png", height=350)

```

```{r}
setVisualStyle("Marquee")
knitr::include_graphics("./demo_marquee.png")
```

```{r}
styles <- getVisualStyleNames()
styles
```


## Read our metagenomics data

We will read in a species co-occurrence matrix that was calculated using Spearman Rank coefficient. (see reference Lima-Mendez et al. (2015) for details).

```{r}
## scripts for processing located in "inst/data-raw/"
prok_vir_cor <- read.delim("virus_prok_cor_abundant.tsv", stringsAsFactors = FALSE)

## Have a peak at the first 6 rows
head(prok_vir_cor)
```

How many edges will we have?

```{r}
nrow(prok_vir_cor)
```

How many unique species are there?

```{r}
length(unique(c(prok_vir_cor[,1], prok_vir_cor[,2])))
```

```{r}
g <-  graph.data.frame(prok_vir_cor, directed = FALSE)
plot(g)
```

```{r}
plot(g, vertex.size = 3, vertex.label = NA)
```

```{r}
library(ggraph)

ggraph(g, layout = 'auto') + 
  geom_edge_link(alpha = 0.25) +
  geom_node_point(color = "steelblue")
```


## Let's calculate some things :)

Vertices and Edges
```{r}
V(g)
E(g)
```

Community detection with the Girvin Newman:

```{r}
cb <-  cluster_edge_betweenness(g)
cb
```

```{r}
plot(cb, y=g, vertex.label=NA,  vertex.size=3)
```

You can extract a cluster/community membership vector for further inspection with the membership() function:

```{r}
head( membership(cb) )
```

Calculate and plot node degree of our network
```{r}

d <- degree(g)
hist(d, breaks=30, col="lightblue", main ="Node Degree Distribution")
```

```{r}
plot( degree_distribution(g), type="h" )
```

## Centrality Analysis

Centrality analysis: Centrality gives an estimation on how important a node or edge is for the connectivity (or the information flow) of a network. It is a particularly useful parameter in signaling networks and it is often used when trying to find drug targets for example.

Centrality analysis often aims to answer the following question: Which nodes are the most important and why?

```{r}
pr <- page_rank(g)
head(pr$vector)
```

Let's plot based on these page rank centrality scores

```{r}
library("BBmisc")
v.size <- BBmisc::normalize(pr$vector, range=c(2,20), method="range")
plot(g, vertex.size=v.size, vertex.label=NA)
```

```{r}
v.size <- BBmisc::normalize(d, range=c(2,20), method="range")
plot(g, vertex.size=v.size, vertex.label=NA)
```

Another very common centrality score is betweenness. The vertex and edge betweenness are (roughly) defined by the number of geodesics (shortest paths) going through a vertex or an edge.

```{r}
b <- betweenness(g)
v.size <- BBmisc::normalize(b, range=c(2,20), method="range")
plot(g, vertex.size=v.size, vertex.label=NA)
```


## Read taxonomic classification for network annotation

```{r}
phage_id_affiliation <- read.delim("phage_ids_with_affiliation.tsv")
head(phage_id_affiliation)
```

```{r}
bac_id_affi <- read.delim("prok_tax_from_silva.tsv", stringsAsFactors = FALSE)
head(bac_id_affi)
```

Extract out our vertex names
```{r}
genenet.nodes <- as.data.frame(vertex.attributes(g), stringsAsFactors=FALSE)
head(genenet.nodes)
```

How many phage entries do we have?
```{r}
length( grep("^ph_",genenet.nodes[,1]) )
```

Now let's merge() these with the annotation data
```{r}
z <- bac_id_affi[,c("Accession_ID", "Kingdom", "Phylum", "Class")]
n <- merge(genenet.nodes, z, by.x="name", by.y="Accession_ID", all.x=TRUE)
head(n)
```

```{r}
colnames(n)
```

```{r}
colnames(phage_id_affiliation)
```

```{r}
# Again we only need a subset of `phage_id_affiliation` for our purposes
y <- phage_id_affiliation[, c("first_sheet.Phage_id_network", "phage_affiliation","Tax_order", "Tax_subfamily")]

# Add the little phage annotation that we have
x <- merge(x=n, y=y, by.x="name", by.y="first_sheet.Phage_id_network", all.x=TRUE)

## Remove duplicates from multiple matches
x <- x[!duplicated( (x$name) ),]
head(x)
```

Save our merged annotation results back to genenet.nodes
```{r}
genenet.nodes <- x
```


## Send network to Cytoscape using RCy3

```{r}
# Open a new connection and delete any existing windows/networks in Cy
deleteAllNetworks()

```

Set the main nodes colname to the required "id"
```{r}
colnames(genenet.nodes)[1] <- "id"
```

```{r}
genenet.edges <- data.frame(igraph::as_edgelist(g))

# Set the main edges colname to the required "source" and "target" 
colnames(genenet.edges) <- c("source","target")

# Add the weight from igraph to a new column...
genenet.edges$Weight <- igraph::edge_attr(g)$weight

# Send as a new network to Cytoscape
createNetworkFromDataFrames(genenet.nodes,genenet.edges, 
                            title="Tara_Oceans")
```




